File: demo/KSYLiveDemo/KSYPlayerDemo/KSYNetTrackerVC.m
//
//  KSYNetworkVC.m
//  KSYPlayerDemo
//
//  Created by 施雪梅 on 2017/1/4.
//  Copyright © 2017年 kingsoft. All rights reserved.
//

#import "KSYUIView.h"
#import "KSYNetTrackerVC.h"

#define ELEMENT_GAP  15

@interface KSYNetTrackerVC() <UITextFieldDelegate>

@end

@implementation KSYNetTrackerVC{
    KSYUIView *ctrlView;
    
    UILabel *lbDomain;
    UITextField *tfDomain;
    UIView * tfDomainLine;
    
    UIButton *btnPing;
    UIButton *btnMTR;
    UIButton *btnQuit;
    
    UITextView *txtView_ret;
    
    KSYNetTracker *tracker;
    BOOL isRunning;
    
    KSY_NETTRACKER_ACTION action;
    NSString *infoLog;
    NSString  *stateStr;
    NSString *displayStr;
    NSMutableArray  *_registeredNotifications;
}

- (void)dealloc {
    [self unregisterObserver];
}

- (void)viewDidLoad {
    [super viewDidLoad];

    [self setupUI];
    //初始化NetTracker
    [self initNetTracker];
    
    infoLog = @"";
    stateStr = @"";
    displayStr = @"";
}

- (void)setupUI {
    ctrlView = [[KSYUIView alloc] initWithFrame:self.view.bounds];
    ctrlView.backgroundColor = [UIColor whiteColor];
    //设置按钮间间距
    ctrlView.gap = ELEMENT_GAP;
    
    @WeakObj(self);
    ctrlView.onBtnBlock = ^(id sender){
        [selfWeak  onBtn:sender];
    };
    //按钮btnPing执行ping命令
    btnPing = [ctrlView addButton:@"Ping"];
    [btnPing setTag:KSY_NETTRACKER_ACTION_PING];
    //按钮btnMTR执行MTR命令
    btnMTR = [ctrlView addButton:@"MTR"];
    [btnMTR setTag:KSY_NETTRACKER_ACTION_MTR];
    //退出按钮
    btnQuit = [ctrlView addButton:@"Quit"];

    lbDomain = [ctrlView addLable:@"请输入待探测地址："];
    lbDomain.backgroundColor = [UIColor whiteColor];
    
    tfDomain = [ctrlView addTextField:@"www.baidu.com"];
    tfDomain.borderStyle = UITextBorderStyleNone;
    tfDomain.returnKeyType = UIReturnKeyDone;
    tfDomain.delegate = self;
    
    tfDomainLine = [ctrlView addLable:nil];
    tfDomainLine.backgroundColor = [UIColor blackColor];
    
    txtView_ret = [[UITextView alloc] initWithFrame:CGRectZero];
    txtView_ret.layer.borderWidth = 1.0f;
    txtView_ret.layer.borderColor = [UIColor lightGrayColor].CGColor;
    txtView_ret.backgroundColor = [UIColor whiteColor];
    txtView_ret.font = [UIFont fontWithName:@"Courier New" size:12.0f];
    txtView_ret.textAlignment = NSTextAlignmentLeft;
    txtView_ret.scrollEnabled = YES;
    txtView_ret.editable = NO;
    [ctrlView addSubview:txtView_ret];
    
    [self layoutUI];
    
    [self.view addSubview: ctrlView];
}

- (void)layoutUI {
    //设置各个空间的fram
    CGFloat wdt = self.view.frame.size.width;
    CGFloat hgt = self.view.frame.size.height;
    int xPos = 0, yPos = 0;
    
    ctrlView.frame = self.view.frame;
    [ctrlView layoutUI];
    
    xPos = wdt / 8;
    yPos = hgt / 15;
    lbDomain.frame = CGRectMake(xPos, yPos, wdt * 4 / 5, ctrlView.btnH);
    
    xPos += 20;
    yPos += ctrlView.btnH + ELEMENT_GAP;
    tfDomain.frame = CGRectMake(xPos, yPos, wdt * 3 / 5, ctrlView.btnH);
    
    tfDomainLine.frame = CGRectMake(xPos ,  yPos + ctrlView.btnH, tfDomain.frame.size.width, 2);
    
    yPos += ctrlView.btnH + 2 * ELEMENT_GAP;
    ctrlView.yPos  = yPos;
    [ctrlView putRow:@[btnPing, btnMTR, btnQuit]];

    yPos += ctrlView.btnH + ELEMENT_GAP;
    txtView_ret.frame = CGRectMake(0, yPos, wdt, hgt - yPos);
}

- (void)onBtn:(UIButton *)btn{
    if (btn == btnPing || btn == btnMTR) {
        //开始探测
        [self startNetDiagnosis: btn];
    }else if (btn == btnQuit){
        [self onQuit];
    }
}

- (void) initNetTracker {
    //创建探测对象KSYNetTracker
    tracker =  [[KSYNetTracker alloc] init];
    if(tracker == nil)
        NSLog(@"init tracker failed\n");
    //监听消息
    [self setupObserver];
}

- (void) setupObserver {
    _registeredNotifications = [[NSMutableArray alloc] init];
    //完成一次探测时收到一次通知
    [[NSNotificationCenter defaultCenter]addObserver:self
                                            selector:@selector(handleTrackerNotify:)
                                                name:(KSYNetTrackerOnceDoneNotification)
                                              object:tracker];
    [_registeredNotifications addObject:KSYNetTrackerOnceDoneNotification];
    //完成所有探测通知时收到此消息
    [[NSNotificationCenter defaultCenter]addObserver:self
                                            selector:@selector(handleTrackerNotify:)
                                                name:(KSYNetTrackerFinishedNotification)
                                              object:tracker];
    [_registeredNotifications addObject:KSYNetTrackerFinishedNotification];
    //探测过程中出现错误时收到此消息
    [[NSNotificationCenter defaultCenter]addObserver:self
                                            selector:@selector(handleTrackerNotify:)
                                                name:(KSYNetTrackerErrorNotification)
                                              object:tracker];
    [_registeredNotifications addObject:KSYNetTrackerErrorNotification];
}

- (void) unregisterObserver {
    //取消消息监听
    for (NSString *name in _registeredNotifications) {
        [[NSNotificationCenter defaultCenter] removeObserver:self
                                                        name:name
                                                      object:tracker];
    }
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

- (void)onQuit {
    displayStr = @"";
    [self displayInfo];
    [self stopNetDiagnosis];
    [self dismissViewControllerAnimated:FALSE completion:nil];
}

- (void)startNetDiagnosis:(UIButton *)button
{
    if(!isRunning){
        displayStr = @"";
        //设置探测方式
        action = button.tag;
        tracker.action = action;
        //开始探测
        if([tracker start:tfDomain.text])
        {
            //启动探测失败
            displayStr = @"启动探测失败，请检查网络或待探测地址!";
            //显示探测结果
            [self displayInfo];
            return ;
        }
        [button setTitle:@"stop" forState:UIControlStateNormal];

        if(action == KSY_NETTRACKER_ACTION_PING)
            btnMTR.enabled = NO;
        else
            btnPing.enabled = NO;
        
        isRunning = !isRunning;
        displayStr  = stateStr = @"开始探测......\n\n";
        [self displayInfo];
    }
    else
    {
        //结束探测
        [self stopNetDiagnosis];
        if(action == KSY_NETTRACKER_ACTION_PING)
            //得到ping探测的结果
            displayStr = [displayStr stringByAppendingString:[self getPingRetStr]];
        else
        {
            displayStr = stateStr = @"停止探测，已统计结果如下：\n";
            displayStr = [displayStr stringByAppendingString:infoLog];
        }
        [self displayInfo];
    }
}

- (void) stopNetDiagnosis
{
    [btnPing setTitle:@"Ping" forState:UIControlStateNormal];
    [btnMTR setTitle:@"MTR" forState:UIControlStateNormal];
    btnMTR.enabled = YES;
    btnPing.enabled = YES;
    //结束探测
    [tracker stop];
    isRunning = NO;
}

- (void)handleTrackerNotify:(NSNotification*)notify
{
    if(tracker == nil)
        return ;
    
    if(KSYNetTrackerOnceDoneNotification == notify.name)
    {
        //完成一次探测时执行下面代码
        if(action == KSY_NETTRACKER_ACTION_PING)
        {
            //本次探测消耗的时间
            float rtt = [[[notify userInfo] valueForKey:@"rtt"] floatValue];
            NSInteger count = [[[notify userInfo] valueForKey:@"count"] integerValue];
            if(rtt < 0.00000001)
                displayStr = [displayStr stringByAppendingFormat:@"Request timeout for icmp_seq %ld\n", count];
            else
            {
                //获取探测结果
                KSYNetRouterInfo  *pingRet = tracker.routerInfo[0];
                displayStr = [displayStr stringByAppendingFormat:@"ping %@ icmp_seq %ld time=%0.3f ms\n", pingRet.ips[0], count, rtt];
            }
        }
        else
        {
            [self getRouterInfo];
            displayStr = @"";
            displayStr = [displayStr stringByAppendingString:stateStr];
            displayStr = [displayStr stringByAppendingString:infoLog];
        }
    }
    else if(KSYNetTrackerFinishedNotification == notify.name)
    {
        //探测完成
        if(action == KSY_NETTRACKER_ACTION_PING)
            displayStr = [displayStr stringByAppendingString:[self getPingRetStr]];
        else
        {
            stateStr = @"探测完成，结果如下：\n\n";
            displayStr = @"";
            displayStr = [displayStr stringByAppendingString:stateStr];
            displayStr = [displayStr stringByAppendingString:infoLog];
        }
        
        [btnPing setTitle:@"Ping" forState:UIControlStateNormal];
        [btnMTR setTitle:@"MTR" forState:UIControlStateNormal];
        btnMTR.enabled = YES;
        btnPing.enabled = YES;
        isRunning = NO;
        [tracker stop];
    }
    else if(KSYNetTrackerErrorNotification == notify.name)
    {
        //探测出现错误
    }
    
    [self displayInfo];
}

- (void) displayInfo{
    //显示探测结果
    dispatch_async(dispatch_get_main_queue(), ^{
        txtView_ret.text = displayStr;
    });
}

- (NSString *) getPingRetStr{
    //返回ping探测的结果
    NSString *pingRetStr= @"";
    KSYNetRouterInfo  *pingRet = tracker.routerInfo[0];
    pingRetStr = [pingRetStr stringByAppendingFormat:@"\n ------ping statics-----\n"];
    pingRetStr = [pingRetStr stringByAppendingFormat:@"%d packets transmitted, %d packets received,  %0.3f packet loss\n",
                  pingRet.number, (int)(pingRet.number * (1 - pingRet.loss)), pingRet.loss];
    
    pingRetStr = [pingRetStr stringByAppendingFormat:@"round-trip min/avg/max/stdev = %0.3f/%0.3f/%0.3f/%0.3fms\n",
                  pingRet.tmin, pingRet.tavg, pingRet.tmax, pingRet.tdev];
    return pingRetStr;
}

- (NSString *) getInfoHeader{
    //包含的各个属性名称
    NSString *header = @"";
    header = [header stringByAppendingFormat:@"%-8s", "idx"];
    header = [header stringByAppendingFormat:@"%-10s", "ip"];
    header = [header stringByAppendingFormat:@"%-8s", "number"];
    header = [header stringByAppendingFormat:@"%-7s", "max"];
    header = [header stringByAppendingFormat:@"%-7s", "min"];
    header = [header stringByAppendingFormat:@"%-6s", "avg"];
    header = [header stringByAppendingFormat:@"%-6s", "stdev"];
    header = [header stringByAppendingFormat:@"%-4s\n", "loss"];
    return header;
}

- (void)getRouterInfo{
    //得到全部统计结果
    infoLog = [self getInfoHeader];
    int i = 1, j = 0;
    for(KSYNetRouterInfo  *netInfo in tracker.routerInfo)
    {
        if(netInfo.ips)
        {
            j = 0;
            for(NSString *ip in netInfo.ips)
            {
                if(j == 0)
                {
                    infoLog = [infoLog stringByAppendingFormat:@"%-3d", i];
                    infoLog = [infoLog stringByAppendingFormat:@"%-16s", [ip UTF8String]];
                    infoLog = [infoLog stringByAppendingFormat:@"%-4d", netInfo.number];
                    infoLog = [infoLog stringByAppendingFormat:@"%5.1fms", netInfo.tmax];
                    infoLog = [infoLog stringByAppendingFormat:@"%5.1fms", netInfo.tmin];
                    infoLog = [infoLog stringByAppendingFormat:@"%5.1fms", netInfo.tavg];
                    infoLog = [infoLog stringByAppendingFormat:@"  %-6.1f", netInfo.tdev];
                    infoLog = [infoLog stringByAppendingFormat:@"%-4.1f\n", netInfo.loss];
                }
                else
                    infoLog = [infoLog stringByAppendingFormat:@"    %-16s\n", [ip UTF8String]];
                j++;
            }
        }
        else
        {
            infoLog = [infoLog stringByAppendingFormat:@"%-3d", i];
            infoLog = [infoLog stringByAppendingFormat:@"%-16s", "--"];
            infoLog = [infoLog stringByAppendingFormat:@"%-4s", "--"];
            infoLog = [infoLog stringByAppendingFormat:@"%-7s", "--"];
            infoLog = [infoLog stringByAppendingFormat:@"%-7s", "--"];
            infoLog = [infoLog stringByAppendingFormat:@"%-7s", "--"];
            infoLog = [infoLog stringByAppendingFormat:@"%-6s", "--"];
            infoLog = [infoLog stringByAppendingFormat:@"%-4s\n\n", "--"];
        }
        i++;
    }
}
-(BOOL)textFieldShouldReturn:(UITextField *)textField
{
    if (textField.returnKeyType == UIReturnKeyDone) {
        [textField resignFirstResponder];
    }
    return YES;
}

@end

File: demo/KSYLiveDemo_Swift/KSYPlayerDemo_Swift/KSYNetTrackerVC.swift
//
//  KSYNetTrackerVC.swift
//  KSYLiveDemo_Swift
//
//  Created by iVermisseDich on 2017/1/23.
//  Copyright © 2017年 com.ksyun. All rights reserved.
//

import UIKit

class KSYNetTrackerVC: UIViewController {

    var lbDomain: UILabel?
    var tfDomain: UITextField?
    var tfDomainLine: UIView?
    
    var btnMTR: UIButton?
    var btnPing: UIButton?
    var btnQuit: UIButton?
    
    var textView_ret: UITextView?
    
    var tracker: KSYNetTracker?
    var isRunning: Bool = false
    var action: KSY_NETTRACKER_ACTION?
    var infoLog: String = ""
    var stateStr: String? = ""
    var displayStr: String? = ""
    var _registeredNotifications: [NSNotification.Name]?
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        initUI()
        initNetTracker()
    }
    
    deinit {
        unregisterObserver()
    }
    
    func initUI() {
        let wdt: CGFloat = view.bounds.width
        let hgt: CGFloat = view.bounds.height
        
        var xPos: CGFloat = 0, yPos: CGFloat = 0
        let elem_width: CGFloat = 100, elem_height: CGFloat = 100
        
        view.backgroundColor = .white
        xPos = wdt / 8
        yPos = hgt / 15
        
        lbDomain = UILabel.init(frame: CGRect.init(x: xPos,
                                                   y: yPos,
                                                   width: wdt * 4 / 5,
                                                   height: elem_height))
        lbDomain?.textColor = .black
        lbDomain?.text = "请输入待探测地址："
        view.addSubview(lbDomain!)
        
        xPos += 20
        yPos += elem_height
        tfDomain = UITextField.init(frame: CGRect.init(x: xPos,
                                                       y: yPos,
                                                       width: wdt * 3 / 5,
                                                       height: elem_height))
        tfDomain?.returnKeyType = .done
        tfDomain?.text = "www.baidu.com"
        
        tfDomainLine = UIView.init(frame: CGRect.init(x: 0,
                                                      y: elem_height,
                                                      width: tfDomain!.frame.width,
                                                      height: 2))
        tfDomainLine?.backgroundColor = .black
        view.addSubview(tfDomainLine!)
        tfDomain?.addSubview(tfDomainLine!)
        
        xPos = wdt / 12
        yPos += elem_height + 25
        btnPing = UIButton.init(type: .roundedRect)
        btnPing?.accessibilityLabel = "Ping"
        btnPing?.frame = CGRect.init(x: xPos,
                                     y: yPos,
                                     width: elem_width,
                                     height: elem_height)
        btnPing?.backgroundColor = .lightGray
        btnPing?.setTitle("Ping", for: .normal)
        btnPing?.addTarget(self, action: #selector(startNetDiagnosis(btn:)), for: .touchUpInside)
        view.addSubview(btnPing!)
        
        xPos += elem_width + 25
        btnMTR = UIButton.init(type: .roundedRect)
        btnMTR?.accessibilityLabel = "MTR"
        btnMTR?.frame = CGRect.init(x: xPos,
                                    y: yPos,
                                    width: elem_width,
                                    height: elem_height)
        btnMTR?.backgroundColor = .lightGray
        btnMTR?.setTitle("MTR", for: .normal)
        btnMTR?.addTarget(self, action: #selector(startNetDiagnosis(btn:)), for: .touchUpInside)
        view.addSubview(btnMTR!)
        
        xPos += elem_width + 25
        btnQuit = UIButton.init(type: .roundedRect)
        btnQuit?.frame = CGRect.init(x: xPos,
                                    y: yPos,
                                    width: elem_width,
                                    height: elem_height)
        btnQuit?.backgroundColor = .lightGray
        btnQuit?.setTitle("btnQuit", for: .normal)
        btnQuit?.addTarget(self, action: #selector(onQuit(sender:)), for: .touchUpInside)
        view.addSubview(btnQuit!)
        
        yPos += elem_height + 25
        textView_ret = UITextView.init(frame: .zero)
        textView_ret?.layer.borderWidth = 1.0
        textView_ret?.layer.borderColor = UIColor.lightGray.cgColor
        textView_ret?.backgroundColor = .white
        textView_ret?.font = UIFont.init(name: "Courier New", size: 12)
        textView_ret?.textAlignment = .left
        textView_ret?.isScrollEnabled = true
        textView_ret?.isEditable = false
        textView_ret?.frame = CGRect.init(x: 0,
                                          y: yPos,
                                          width: wdt,
                                          height: hgt - yPos)
        view.addSubview(textView_ret!)
        
    }
    
    func initNetTracker() {
        tracker = KSYNetTracker()
        if let _ = tracker {
            print("init tracker failed")
        }
        setupObserver()
    }
    
    func setupObserver() {
        
        NotificationCenter.default.addObserver(self, selector: #selector(handleTrackerNotify(notify:)), name: NSNotification.Name.KSYNetTrackerOnceDone, object: tracker)
        
        NotificationCenter.default.addObserver(self, selector: #selector(handleTrackerNotify(notify:)), name: NSNotification.Name.KSYNetTrackerFinished, object: tracker)
        
        NotificationCenter.default.addObserver(self, selector: #selector(handleTrackerNotify(notify:)), name: NSNotification.Name.KSYNetTrackerError, object: tracker)
        _registeredNotifications = [.KSYNetTrackerOnceDone, .KSYNetTrackerFinished, .KSYNetTrackerError]
    }
    
    func unregisterObserver() {
        guard let _ = _registeredNotifications else {
            NotificationCenter.default.removeObserver(self)
            return
        }
        
        for name in _registeredNotifications! {
            NotificationCenter.default.removeObserver(self, name: name, object: tracker)
        }
    }
    
    func onQuit(sender: UIButton) {
        displayStr = ""
        displayInfo()
        stopNetDiagnosis()
        dismiss(animated: false, completion: nil)
    }
    
    func startNetDiagnosis(btn: UIButton) {
        if !isRunning {
            displayStr = ""
            
            if btn.accessibilityLabel == "Ping" {
                action = KSY_NETTRACKER_ACTION.PING
            }else{
                action = KSY_NETTRACKER_ACTION.MTR
            }
            
            tracker?.action = action!
            if let _ = tracker ,let _ = tfDomain{
                if tracker!.start((tfDomain!.text)!) != 0 {
                    displayStr = "启动探测失败，请检查网络或待探测地址!"
                    displayInfo()
                    return
                }
            }
            
            btn.setTitle("stop", for: .normal)
            if btn.accessibilityLabel == "Ping" {
                btnMTR?.alpha = 0.4
                btnMTR?.isEnabled = false
            }else {
                btnPing?.alpha = 0.4
                btnPing?.isEnabled = false
            }
            
            isRunning = !isRunning
            displayStr = "开始探测......\n"
            stateStr = "开始探测......\n"
            displayInfo()
        }else{
            stopNetDiagnosis()
            if action == KSY_NETTRACKER_ACTION.PING {
                displayStr = displayStr?.appending(getPingRetStr())
            }else{
                displayStr = "停止探测，已统计结果如下：\n"
                stateStr = "停止探测，已统计结果如下：\n"
                displayStr = displayStr?.appending(infoLog ?? "")
            }
            displayInfo()
        }
    }
    
    func stopNetDiagnosis() {
        btnPing?.setTitle("Ping", for: .normal)
        btnMTR?.setTitle("MTR", for: .normal)
        btnMTR?.alpha = 1
        btnPing?.alpha = 1
        btnMTR?.isEnabled = true
        btnPing?.isEnabled = true
        
        tracker?.stop()
        isRunning = false
    }
    
    func handleTrackerNotify(notify: NSNotification) {
        guard let _ = tracker else {
            return
        }
        
        switch notify.name {
        case Notification.Name.KSYNetTrackerOnceDone:
            if action == KSY_NETTRACKER_ACTION.PING {
                let rtt = ((notify.userInfo! as NSDictionary).object(forKey: "rtt") as! NSString).floatValue
                let count = ((notify.userInfo! as NSDictionary).object(forKey: "count") as! NSString).intValue
                if rtt < 0.00000001 {
                    displayStr = displayStr?.appending("Request timeout for icmp_seq \(count)")
                }else{
                    let pingRet = tracker?.routerInfo?.firstObject as! KSYNetRouterInfo
                    displayStr = displayStr?.appending(String(format: "ping \(pingRet.ips?.firstObject ?? "") icmp_seq \(count) time=%0.3f ms", rtt))
                }
            }else {
                getRouterInfo()
                displayStr = ""
                displayStr = displayStr?.appending(stateStr ?? "")
                displayStr = displayStr?.appending(infoLog ?? "")
            }
            break
        case Notification.Name.KSYNetTrackerFinished:
            if action == KSY_NETTRACKER_ACTION.PING {
                displayStr = displayStr?.appending(getPingRetStr())
            }else {
                stateStr = "探测完成，结果如下：\n\n"
                displayStr = ""
                displayStr = displayStr?.appending(stateStr ?? "")
                displayStr = displayStr?.appending(infoLog ?? "")
            }
            
            btnPing?.setTitle("Ping", for: .normal)
            btnMTR?.setTitle("MTR", for: .normal)
            btnMTR?.alpha = 1
            btnPing?.alpha = 1
            btnMTR?.isEnabled = true
            btnPing?.isEnabled = true
            isRunning = false
            tracker?.stop()
            break
        case Notification.Name.KSYNetTrackerError:
            break
        default:
            ()
        }
        displayInfo()
    }
    
    func displayInfo() {
        DispatchQueue.main.async { [weak self] in
            self?.textView_ret?.text = self?.displayStr
        }
    }
    
    func getPingRetStr() -> String {
        var pingRetStr: String = ""
        let pingRet = tracker?.routerInfo?[0] as! KSYNetRouterInfo
        
        pingRetStr = pingRetStr.appending("\n ------ping statics-----\n")
        pingRetStr = pingRetStr.appendingFormat("%d packets transmitted, %d packets received,  %0.3f packet loss\n", pingRet.number, Int(Float(pingRet.number) * (1 - pingRet.loss)), pingRet.loss)
        
        pingRetStr = pingRetStr.appendingFormat("round-trip min/avg/max/stdev = %0.3f/%0.3f/%0.3f/%0.3fms\n", pingRet.tmin, pingRet.tavg, pingRet.tmax, pingRet.tdev)
        return pingRetStr
    }
    
    func getInfoHeader() -> String {
        var header = ""
        header = header.appendingFormat("%-8s", "idx")
        header = header.appendingFormat("%-10s", "ip")
        header = header.appendingFormat("%-8s", "number")
        header = header.appendingFormat("%-7s", "max")
        header = header.appendingFormat("%-7s", "min")
        header = header.appendingFormat("%-6s", "avg")
        header = header.appendingFormat("%-6s", "stdev")
        header = header.appendingFormat("%-4s\n", "loss")
        return header
    }
    
    func getRouterInfo() {
        infoLog = getInfoHeader()
        var i = 1, j = 0
        
        guard let _ = tracker?.routerInfo else {
            return
        }
        
        tracker!.routerInfo?.enumerateObjects( { (netInfo, idx, _) in
            if let info: KSYNetRouterInfo = (netInfo as! KSYNetRouterInfo) {
                
                if let _ = info.ips {
                    j = 0;
                    
                    for ip in info.ips! {
                        if let ip: String = (ip as! String) {
                            if j == 0 {
                                infoLog = infoLog.appendingFormat("%-3d", i)
                                infoLog = infoLog.appendingFormat("%-16s", (ip as NSString).utf8String!)
                                infoLog = infoLog.appendingFormat("%-4d", info.number)
                                infoLog = infoLog.appendingFormat("%5.1fms", info.tmax)
                                infoLog = infoLog.appendingFormat("%5.1fms", info.tmin)
                                infoLog = infoLog.appendingFormat("%5.1fms", info.tavg)
                                infoLog = infoLog.appendingFormat("%  %-6.1f", info.tdev)
                                infoLog = infoLog.appendingFormat("%-4.1f\n", info.loss)
                            }else{
                                infoLog = infoLog.appendingFormat("    %-16s\n", ip)
                            }
                            
                            j += 1
                        }
                    }
                }else{
                    infoLog = infoLog.appendingFormat("%-3d", i)
                    infoLog = infoLog.appendingFormat("%-16s", "--")
                    infoLog = infoLog.appendingFormat("%-4s", "--")
                    infoLog = infoLog.appendingFormat("%-7s", "--")
                    infoLog = infoLog.appendingFormat("%-7s", "--")
                    infoLog = infoLog.appendingFormat("%-7s", "--")
                    infoLog = infoLog.appendingFormat("%-6s", "--")
                    infoLog = infoLog.appendingFormat("%-4s\n\n", "--")
                }
                i += 1
            }
        });
        
//        for netInfo in tracker!.routerInfo {
//            if let info: KSYNetRouterInfo = (netInfo as! KSYNetRouterInfo) {
//                
//                if let _ = info.ips {
//                    j = 0;
//                    
//                    for ip in info.ips {
//                        if let ip: String = (ip as! String) {
//                            if j == 0 {
//                                infoLog = infoLog.appendingFormat("%-3d", i)
//                                infoLog = infoLog.appendingFormat("%-16s", ip)
//                                infoLog = infoLog.appendingFormat("%-4d", info.number)
//                                infoLog = infoLog.appendingFormat("%5.1fms", info.tmax)
//                                infoLog = infoLog.appendingFormat("%5.1fms", info.tmin)
//                                infoLog = infoLog.appendingFormat("%5.1fms", info.tavg)
//                                infoLog = infoLog.appendingFormat("%  %-6.1f", info.tdev)
//                                infoLog = infoLog.appendingFormat("%-4.1f\n", info.loss)
//                            }else{
//                                infoLog = infoLog.appendingFormat("    %-16s\n", ip)
//                            }
//                            
//                            j += 1
//                        }
//                    }
//                }else{
//                    infoLog = infoLog.appendingFormat("%-3d", i)
//                    infoLog = infoLog.appendingFormat("%-16s", "--")
//                    infoLog = infoLog.appendingFormat("%-4s", "--")
//                    infoLog = infoLog.appendingFormat("%-7s", "--")
//                    infoLog = infoLog.appendingFormat("%-7s", "--")
//                    infoLog = infoLog.appendingFormat("%-7s", "--")
//                    infoLog = infoLog.appendingFormat("%-6s", "--")
//                    infoLog = infoLog.appendingFormat("%-4s\n\n", "--")
//                }
//                i += 1
//            }
//        }
    }
    
}
